src/analysis/mesh/qgsmeshcalcnode.h

```diff
@@ -0,0 +1,164 @@
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSMESHCALCNODE_H
+#define QGSMESHCALCNODE_H
+
+#define SIP_NO_FILE
+
+///@cond PRIVATE
+
+#include <QMap>
+#include <QString>
+#include <QStringList>
+#include <memory>
+#include <limits>
+
+#include "qgis_analysis.h"
+
+#include "qgsmeshcalcutils.h"
+
+struct QgsMeshMemoryDatasetGroup;
+
+/**
+ * \ingroup analysis
+ * \class QgsMeshCalcNode
+ * Represents a single calculation node
+ *
+ * \since QGIS 3.6
+ */
+class ANALYSIS_EXPORT QgsMeshCalcNode
+{
+  public:
+    //! types of mesh node
+    enum Type
+    {
+      tOperator = 1, //!< Operator (e.g. +, -)
+      tNumber, //!< Number (e.g. 1)
+      tNoData, //!< Nodata (NaN)
+      tDatasetGroupRef //!< Dataset group
+    };
+
+    //! operators between dataset groups
+    enum Operator
+    {
+      opPLUS,     //!< Plus
+      opMINUS,    //!< Minus
+      opMUL,      //!< Multiply
+      opDIV,      //!< Divide
+      opPOW,      //!< Power
+      opEQ,       //!< Equal
+      opNE,       //!< Not equal
+      opGT,       //!< Greater than
+      opLT,       //!< Lower than
+      opGE,       //!< Greater or equal
+      opLE,       //!< Lower or equal
+      opAND,      //!< Boolean AND
+      opOR,       //!< Boolean OR
+      opNOT,      //!< Boolean NOT (unary)
+      opIF,       //!< Boolean IF
+      opSIGN,     //!< Change sign (unary)
+      opMIN,      //!< Minimum value
+      opMAX,      //!< Maximum value
+      opABS,      //!< Absolute value (unary)
+      opSUM_AGGR, //!< Aggregated sum (unary)
+      opMAX_AGGR, //!< Aggregated maximum (unary)
+      opMIN_AGGR, //!< Aggregated minimum (unary)
+      opAVG_AGGR, //!< Aggregated average (unary)
+      opNONE,     //!< Nodata value
+    };
+
+    /**
+     * Constructs a Type::tNoData node initialized with NODATA values
+     */
+    QgsMeshCalcNode();
+
+    /**
+     * Constructs a Type::tNumber node initialized with selected number
+     */
+    QgsMeshCalcNode( double number );
+
+    /**
+     * Constructs a Type::tOperator node for binary or unary operator
+     * \param op Operator to perform
+     * \param left Left node. This node takes ownership of the node
+     * \param right Right node (for binary operators). This node takes ownership of the node
+     */
+    QgsMeshCalcNode( Operator op, QgsMeshCalcNode *left, QgsMeshCalcNode *right );
```

Comment:
the file has `#define SIP_NO_FILE`, in private API for doxygen and is exported only because of unittest.

Reply:
In that case, could you make the argument a unique_ptr instead of the raw pointer?

Reply:
I can but I am not sure how easy it is. These functions are used in qgsmeshcalcparser.yy and that file is copy-paste of raster equivalent. And since this is preprocessed file I would rather not modify it too much...

Reply:
Ok, that's definitely a good move then! Spend as little time in that parser code as you possibly can :stuck_out_tongue: